File: stack/735_AstoriodCollision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < asteroids.length; i++) {
            int newStone = asteroids[i];
            while (!stack.isEmpty() && stack.peek() > 0 && newStone < 0 && Math.abs(stack.peek()) < Math.abs(newStone)) {
                stack.pop();
            }
            if (stack.isEmpty()) {
                stack.push(newStone);
            } else {
                if (stack.peek() > 0 && newStone < 0) {
                    if (Math.abs(stack.peek()) == Math.abs(newStone)) {
                        stack.pop();
                    }
                } else {
                    stack.push(newStone);
                }
            }
        }
        int[] output = new int[stack.size()];
        int index = output.length - 1;
        while (!stack.isEmpty()) {
            output[index--] = stack.pop();
        }
        return output;
    }
}

File: stack/735_AstoriodCollision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for asteroid in asteroids:
            if len(stack) == 0:
                stack.append(asteroid)
            else:
                if asteroid < 0:
                    asteroid_destroyed = False
                    while len(stack) > 0 and stack[-1] > 0:
                        if abs(stack[-1]) > abs(asteroid):
                            asteroid_destroyed = True
                            break
                        elif abs(stack[-1]) < abs(asteroid):
                            stack.pop()
                        else:
                            stack.pop()
                            asteroid_destroyed = True
                            break
                    if not asteroid_destroyed:
                        stack.append(asteroid)
                else:
                    stack.append(asteroid)
        return stack